### PR TITLE
add event for inserting cache entries

### PR DESCRIPTION
**Pull request by icewind1991**
this provides a reliable way for apps to listen to new files
without the need to of cache wrappers to hook into inserts themselves
(something which isn't 100% reliable)

### CODE REVIEW

File: lib/private/Files/Cache/Cache.php [hunk 1]
```diff
@@ -283,9 +286,11 @@ public function insert($file, array $data) {
 			}
 
 			if ($builder->execute()) {
-				return (int)$this->connection->lastInsertId('*PREFIX*filecache');
+				$fileId = (int)$this->connection->lastInsertId('*PREFIX*filecache');
+				$this->eventDispatcher->dispatch('OC\Files\Cache\Cache::insert', new CacheInsertEvent($this->storage, $file, $fileId));
```
Comment (rullzer) [inline lib/private/Files/Cache/Cache.php]: Would it not be cleaner to have a const in ICache? Then it is public.

So in ICache something like

`const INSERTEVENT = self::class . '::insert'`

Then everybody can just use that event to listen to which is then in OCP?

Comment (rullzer) [reply lib/private/Files/Cache/Cache.php]: Ah just use `CacheInsertEvent::class`?

File: lib/private/Files/Cache/CacheInsertEvent.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OC\Files\Cache;
```
Comment (rullzer) [inline lib/private/Files/Cache/CacheInsertEvent.php]: Should this not be in OCP for apps to use it?

Comment (rullzer) [reply lib/private/Files/Cache/CacheInsertEvent.php]: Or is all the other stuff also in OC? Because then I'm fine

**Comment by icewind1991** [comment]
@rullzer all done

**Review by rullzer — APPROVED** [review]
I like it!

**Review by MorrisJobke — APPROVED** [review]
Code looks good and works 👍

**Pull request merged**